In [1]:
# librerias
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np
import pandas as pd

#
import functions as f


In [2]:
#load the data
df_train= pd.read_csv("Data/ParisHousing_train_data_.csv")
df_test= pd.read_csv("Data/ParisHousing_test_data.csv")
df_val= pd.read_csv("Data/ParisHousing_val_data.csv")


#split
df_train_x =df_train.drop(columns=['price'])  
df_train_y=df_train['price'] 

df_test_x= df_test.drop(columns=['price'])  
df_test_y = df_test['price'] 

df_val_x= df_val.drop(columns=['price'])  
df_val_y= df_val['price'] 

In [3]:
# create model
mlr_model = LinearRegression()

# fit the model
mlr_model.fit(df_train_x, df_train_y)

# make predictions
y_pred = mlr_model.predict(df_test_x)


In [4]:
# Evaluate the model
mse = mean_squared_error(df_test_y, y_pred)
mae = mean_absolute_error(df_test_y, y_pred)
mape = f.mean_absolute_percentage_error(df_test_y, y_pred)  
r2 = r2_score(df_test_y, y_pred)          

# results
print("Model coefficients:", mlr_model.coef_)
print("Intercept:", mlr_model.intercept_)
print("MSE (Mean Squared Error):", mse)
print("MAE (Mean Absolute Error):", mae)
print("MAPE (Mean Absolute Percentage Error):", mape, "%")
print("Determination coefficient (R²):", r2)

Model coefficients: [ 2.86170756e+06  2.22091276e+01  1.49839212e+03  1.48536770e+03
  1.57778914e+03 -1.03442943e+01  1.55366914e+02  1.12099880e+01
 -2.42262476e+01  1.00319224e+02  7.71279138e+01 -1.44575535e+01
 -5.50697420e+00  2.58250912e+01 -6.36116792e-01 -4.49043145e+00]
Intercept: 4992637.822749999
MSE (Mean Squared Error): 7453887.663190103
MAE (Mean Absolute Error): 2190.698205351372
MAPE (Mean Absolute Percentage Error): 0.13044161572912427 %
Determination coefficient (R²): 0.9999991342883362


## **Model Coefficients**
- Each coefficient represents the weight of a feature in predicting the target variable (`price`).
- For example, a coefficient of `1.49839212e+03` for `hasPool` indicates that having a pool increases the predicted price by approximately 1,498 units (holding all other features constant).

## **Model Evaluation**
- **MSE (Mean Squared Error):** `3433482.6252788133`
  - Represents the average squared error between predicted and actual prices. Smaller values indicate better performance.
- **MAE (Mean Absolute Error):** `1434.7351100877852`
  - Measures the average absolute error. On average, predictions deviate by ~1435 units from actual prices.
- **MAPE (Mean Absolute Percentage Error):** `0.11455074798610643 %`
  - Indicates the percentage error. The model is very accurate, with predictions deviating by only ~0.11% on average.
- **R² (Determination Coefficient):** `0.9999996012274278`
  - Almost perfect fit, meaning the model explains ~100% of the variance in the data.


## **Summary of Results**
1. **High Accuracy**: 
   - The model performs exceptionally well, with minimal errors (low MSE, MAE, and MAPE) and an almost perfect R² value.

2. **Feature Insights**:
   - Some features (e.g., `squareMeters` or `hasPool`) have large coefficients, indicating a significant impact on price.
   - Others (e.g., `hasStorageRoom`, `hasGuestRoom`) have smaller coefficients, suggesting lesser influence.

3. **Baseline Price**:
   - The intercept of `4992637.822749999` suggests that properties without additional features still have a high baseline price.
